### PR TITLE
Use sorted witnesses to reduce gas spend

### PR DESCRIPTION
**Pull request by banshee**

### CODE REVIEW

File: smart-contracts/contracts/CosmosBridge.sol [hunk 1]
```diff
@@ -222,13 +223,19 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
...
 
-      for (uint256 j = i + 1; j < validatorLength;) {
-        require(validator.signer != _validators[j].signer, "DUP_SIGNER");
-        unchecked { ++j; }
+      // Signatures must be sorted on the sifchain side, so we just
+      // need to make sure that the next validator in the array
+      // (if we're not at the end) isn't a duplicate and is
+      // sorted correctly
+      if (i + 1 <= validatorLength - 1) {
+        if (validator.signer >= _validators[i + 1].signer) {
```
Comment (Brando753) [inline smart-contracts/contracts/CosmosBridge.sol:234]: This cannot be >= it must be > or else duplicate signatures could get through right?

File: x/instrumentation/test_instrumentation.go [hunk 2]
```diff
@@ -44,9 +47,9 @@ const (
 )
 
 func PeggyCheckpoint(logger log.Logger, kind string, keysAndValues ...interface{}) {
-	logger.Debug(peggyTestMarker, append([]interface{}{kindMarker, kind}, keysAndValues...)...)
+	logger.Debug(peggyTestMarker, append([]interface{}{kindMarker, kind, "cmdline", strings.Join(os.Args, " ")}, keysAndValues...)...)
```
Comment (Brando753) [inline x/instrumentation/test_instrumentation.go:50]: Is this just logging the command line arguments out as a debug line?

**Review by Brando753 — CHANGES_REQUESTED** [review]

**Review by Brando753 — APPROVED** [review]

File: cmd/ebrelayer/txs/relayToEthereum.go [hunk 3]
```diff
@@ -78,6 +80,85 @@ func InitRelayConfig(
 	return client, transactOptsAuth, target, nil
 }
 
+type BatchUnit struct {
```
[the remark this answers is not in the document]
Comment (banshee) [reply cmd/ebrelayer/txs/relayToEthereum.go]: Done

File: x/instrumentation/test_instrumentation.go (hunk 2 above)
Comment (banshee) [reply x/instrumentation/test_instrumentation.go:50]: Yes

**Pull request merged**